indra/newview/llterrainpaintmap.cpp

```diff
@@ -283,3 +301,690 @@ bool LLTerrainPaintMap::bakeHeightNoiseIntoPBRPaintMapRGB(const LLViewerRegion&
+        // Because the paint data may not cover a whole power-of-two region,
+        // allocate a bigger 2x2 image if needed, but set the image data later
+        // for a subset of the image.
+        // Pixel data outside this subset is left undefined. We will use
+        // TERRAIN_STAMP_SCALE in the stamp shader to define the subset of the
+        // image we care about.
+        const U32 width_rounded =  1 << U32(ceil(log2(F32(paint_in->mWidthX))));
+        const U32 height_rounded = 1 << U32(ceil(log2(F32(paint_in->mWidthY))));
+        LLPointer<LLImageGL> stamp_image;
+        {
+            // Create image object (dimensions not yet initialized in GL)
+            U32 stamp_tex_name;
+            LLImageGL::generateTextures(1, &stamp_tex_name);
+            const U32 components = paint_in->mComponents;
+            constexpr LLGLenum target = GL_TEXTURE_2D;
+            const LLGLenum internal_format = paint_in->mComponents == 4 ? GL_RGBA8 : GL_RGB8;
+            const LLGLenum format = paint_in->mComponents == 4 ? GL_RGBA : GL_RGB;
+            constexpr LLGLenum type = GL_UNSIGNED_BYTE;
+            stamp_image = new LLImageGL(stamp_tex_name, components, target, internal_format, format, type, LLTexUnit::TAM_WRAP);
+            // Nearest-neighbor filtering to reduce surprises
+            stamp_image->setFilteringOption(LLTexUnit::TFO_POINT);
+
+            // Initialize the image dimensions in GL
+            constexpr U8* undefined_data_for_now = nullptr;
+            gGL.getTexUnit(0)->bind(stamp_image, false, true);
+            glTexImage2D(GL_TEXTURE_2D, 0, internal_format, width_rounded, height_rounded, 0, format, type, undefined_data_for_now);
+            gGL.getTexUnit(0)->unbind(LLTexUnit::TT_TEXTURE);
+            stamp_image->setSize(width_rounded, height_rounded, components);
+            stamp_image->setDiscardLevel(0);
+
+            // Manually set a subset of the image in GL
+            const U8* data = paint_in->mData.data();
+            const S32 data_width = paint_in->mWidthX;
+            const S32 data_height = paint_in->mWidthY;
+            constexpr S32 origin = 0;
+            // width = data_width; height = data_height. i.e.: Copy the full
+            // contents of data into the image.
+            stamp_image->setSubImage(data, data_width, data_height, origin, origin, /*width=*/data_width, /*height=*/data_height);
+        }
+
+        // Apply ("stamp") the paint to the render target
+        {
+            shader.bindTextureImageGL(LLShaderMgr::DIFFUSE_MAP, stamp_image);
+            const F32 width_fraction = F32(paint_in->mWidthX) / F32(width_rounded);
+            const F32 height_fraction = F32(paint_in->mWidthY) / F32(height_rounded);
+            shader.uniform2f(LLShaderMgr::TERRAIN_STAMP_SCALE, width_fraction, height_fraction);
+            buf->setBuffer();
+            buf->draw(LLRender::TRIANGLES, buf->getIndicesSize(), 0);
+        }
+
```

Comment:
I realize this is a preliminary version of this, but is here where we would want to possibly return and enqueue a job to later check the results via `glReadPixels()` to avoid stalling?

Reply:
The intent of the queues is to allow them to be called somewhere more sensible than directly in llviewermenu.cpp . I'm not 100% sure where they would land, but I would consider the multithreading limitations of OpenGL, plus the desire for real-time painting.